### PR TITLE
Add support for curly brackets in SnowSQL ampersand variables

### PR DESCRIPTION
**Pull request by chwiese**
### Brief summary of the change made

Moves support for Snowsql variables from the dialect into the placeholder templater and adds support for curly brackets, as documented [here](https://docs.snowflake.com/en/user-guide/snowsql-use.html#substituting-variables-in-a-session). Since Snowsql variables can hold SQL logic such as optional WHERE clauses, the templater is the better home for them as discussed [here](https://sqlfluff.slack.com/archives/C01JS7H598E/p1636892634079800?thread_ts=1636886949.072400&cid=C01JS7H598E) on Slack.

### Are there any other side effects of this change that we should be aware of?

Yes, anybody how currently has Snowsql variables in their scripts will run into parsing errors. They will need to switch to the placeholder templater and define the placeholders in their Sqlfluff config.

### Pull Request checklist
- [X] Please confirm you have completed any of the necessary steps below.

- Included test cases to demonstrate any code changes, which may be one or more of the following:
  - `.yml` rule test cases in `test/fixtures/rules/std_rule_cases`.
  - `.sql`/`.yml` parser test cases in `test/fixtures/dialects` (note YML files can be auto generated with `python test/generate_parse_fixture_yml.py` or by running `tox` locally).
  - Full autofix test cases in `test/fixtures/linter/autofix`.
  - Other.
- Added appropriate documentation for the change.
- Created GitHub issues for any relevant followup/future enhancements if appropriate.

### CODE REVIEW

File: src/sqlfluff/dialects/dialect_snowflake.py [hunk 1]
```diff
@@ -105,7 +107,7 @@
         type="variable",
     ),
     ReferencedVariableNameSegment=RegexParser(
-        r"(\$|\&)[A-Z][A-Z0-9_]*",
+        r'(?:\$[A-Z][A-Z0-9_]*)|(?:[A-Z0-9_\."]*&{[A-Z0-9_]*}[A-Z0-9_]*)|(?:&[A-Z0-9_\."]*)',
```
Comment (chwiese) [inline src/sqlfluff/dialects/dialect_snowflake.py:110]: For some reason `(?:[\w\."]*&{[A-Z0-9_]*}[\w]*)` does not work, while `(?:[A-Z0-9_\."]*&{[A-Z0-9_]*}[A-Z0-9_]*)` does. Does anybody know why?

Comment (tunetheweb) [reply src/sqlfluff/dialects/dialect_snowflake.py:110]: Those should be identical. Can you define "does not work"? Have you been able to reproduce it on https://regex101.com/?

Comment (chwiese) [reply src/sqlfluff/dialects/dialect_snowflake.py:110]: Sorry, you are right, "does not work" is rarely helpful.
When I use `r'(?:\$[A-Z][A-Z0-9_]*)|(?:[A-Z0-9_\."]*&{[A-Z0-9_]*}[A-Z0-9_]*)|(?:&[A-Z0-9_\."]*)',`, then `USE WAREHOUSE &WAREHOUSE;` parses alright. If I use `r'(?:\$[A-Z][A-Z0-9_]*)|(?:[\w\."]*&{[A-Z0-9_]*}[A-Z0-9_]*)|(?:&[\w\."]*)',`, the same piece of SQL returns a _Found unparsable section_ error.
I cannot reproduce this on regex101.com - over there both patterns produce the desired match `&WAREHOUSE`.

Comment (WittierDinosaur) [reply src/sqlfluff/dialects/dialect_snowflake.py:110]: This is a bug - not in your regex - but in the RegexParser that is being used. RegexParser inherits from StringParser, which calls this in its __init__ (which is called in the RegexParser)
```
self.template = template.upper()
```
So, `\w` becomes `\W`, which matches every character that is not a word character. Basically `^[\w]`

**Review by barrywhart — APPROVED** [review]
The code changes look good to me. I do have a question: This PR appears to add support for ampersand variables both in the placeholder templater as well as the Snowflake dialect. Do both of them use the same syntax? Is it possible to use both at once? Do we need to add or update documentation so users will understand what's supported and how to use it?

**Comment by chwiese** [comment]
@barrywhart Actually this PR removes ampersand support from the Snowflake dialect and moves it to the placeholder templater only. There is still dollar sign variable support in the dialect though.

Both can be used concurrently. Dollar sign variables are part of the dialect and can be used with all clients, ampersand variables are specific to the official SnowSQL client.

**Comment by barrywhart** [comment]
Got it! Merging now. Great work!

**Pull request merged**